src/main/java/com/github/princesslana/eriscasper/ErisCasper.java

```diff
@@ -26,41 +30,107 @@
   private final ObjectMapper jackson = Jackson.newObjectMapper();
   private final Payloads payloads = new Payloads(jackson);
 
+  private final Integer[] shard;
+
   private final Routes routes;
 
-  private ErisCasper(BotToken token) {
-    this.token = token;
+  private Gateway gateway;
 
+  private ErisCasper(BotToken token, Integer[] shard) {
+    this.token = token;
+    this.shard = shard;
     routes = new Routes(token, httpClient, jackson);
   }
 
   private Observable<Event> getEvents() {
-    Gateway gateway = Gateway.create(httpClient, payloads);
-
+    gateway = Gateway.create(httpClient, payloads);
     return Single.just(RouteCatalog.getGateway())
         .observeOn(Schedulers.io())
         .flatMap(routes::execute)
         .toObservable()
-        .flatMap(gr -> gateway.connect(gr.getUrl(), token))
+        .flatMap(gr -> gateway.connect(gr.getUrl(), token, shard))
         .observeOn(Schedulers.computation())
         .share();
   }
 
+  public Integer getShardNum() {
+    return shard == null ? -1 : shard[0];
+  }
+
   public void run(Bot bot) {
     Observable<Event> events = getEvents();
 
     RepositoryManager rm = RepositoryManager.create(events);
-
-    bot.apply(new BotContext(events, routes, rm))
+    bot.apply(new BotContext(this, events, routes, rm))
         .doOnError(t -> LOG.warn("Exception thrown by Bot", t))
         .blockingAwait();
   }
 
+  public void shutdownGracefully() {
+    gateway.shutdownGracefully();
+  }
+
+  public Completable requestGuildMembers(RequestGuildMembers request) {
```

Comment:
This looks like a candidate for a repository somewhere.

When the client wants to fetch data, they should go via repositories. When they want to do something, it should be an Action.

Reply:
This isn't fetching members. This is asking for offline guild members to be noticed in very large guilds. The naming is a bit weird for this.

Reply:
ok, it would still seem to be better in a repository as `getOfflineGuildMembers`.

Reply:
Maybe but I'd much rather support it the same as all the other gateway commands, it doesn't return anything, it just sends `GuildMemberChunkEvent`s so it's not like it returns anything, it just sends a request for events to update guild members. We could throw some sort of implementation in the repo.